### PR TITLE
[fix] updating solve and factor_plu

### PR DESCRIPTION
**Pull request by christalee**
Changes made to `factor_plu()` focus on matching the CLRS-given pseudocode, specifically un-nesting ln 51-53 and adding ln 60-62. Changes made to `solve()` affect the loops in ln 164 and 166, changing the end values to `N+C`. All other changes are whitespace.

### CODE REVIEW

File: solve.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    if (p == 0) {
+      throw "singular matrix";
+    }
+    p = permutations[k];
+    permutations[k] = permutations[k_swap];
+    permutations[k_swap] = p;
+
+    // elimination
+    for (let i = 0; i < N; i++) {
+      let a = A[k][i];
```
Comment (bergey) [inline solve.ts]: - `const a`
- comment "elimination" should go on the next block, not this one, which swaps rows

**Review by bergey — COMMENTED** [review]
Why move all code to a new file?  If this is deliberate, please update `package.json` to match.

**Comment by christalee** [comment]
I wanted a more descriptive name than `index` when I started writing tests & importing from it, but if that is the proper name for the entrypoint I will change it back. It will make the diff easier to read, and we can always change it later.

File: index.ts [hunk 2]
```diff
@@ -1,192 +1,198 @@
...
-        let p = 0;
-        let k_swap = k;
-        for (let i = k; k < N; i++) {
-            if (Math.abs(A[i][k]) > p) {
-                p = Math.abs(A[i][k]);
-                k_swap = i;
-            }
-            if (p == 0) {
-                throw "singular matrix";
-            }
```
Comment (christalee) [inline index.ts:52]: I pulled this clause out of the loop, based on the implementation given in CLRS.

File: index.ts [hunk 3]
```diff
@@ -1,192 +1,198 @@
...
+    p = permutations[k];
+    permutations[k] = permutations[k_swap];
+    permutations[k_swap] = p;
 
-    const permutations: number[] = [];
     for (let i = 0; i < N; i++) {
-        permutations.push(i);
+      const a = A[k][i];
+      A[k][i] = A[k_swap][i];
+      A[k_swap][i] = a;
```
Comment (christalee) [inline index.ts:61]: I added this section based on the implementation give in CLRS.

File: index.ts [hunk 4]
```diff
@@ -1,192 +1,198 @@
...
+  for (const c of components) {
+    N = Math.max(N, c.p, c.q);
+  }
+
+  // square matrix, all zeros, one row (equation) per component, 1 per node except node 0
+  const A: number[][] = [];
+  const b: number[] = [];
+  for (let i = 0; i < N + C; i++) {
+    const row: number[] = [];
+    for (let j = 0; j < N + C; j++) {
```
Comment (christalee) [inline index.ts:166]: I made changes here to ensure A is a square matrix.

**Review by bergey — APPROVED** [review]
I agree the `index.ts` convention is not great.  I just went with it because `yarn init` set it up that way.  But this diff is much easier to read, and a rename in a separate commit will be easy to read.

**Pull request merged**